addons/knowledge/models/knowledge_article.py

```diff
@@ -0,0 +1,992 @@
+
+    @api.returns('self', lambda value: value.id)
+    def copy(self, default=None):
+        self = self.sudo()
+        default = dict(default or {},
+                       name=_("%s (copy)", self.name),
+                       sequence=self.sequence+1)
+        return super().copy(default=default)
+
+    def unlink(self):
+        for article in self:
+            # Make all the article's children be adopted by the parent's parent.
+            # Otherwise, we will have to manage an orphan house.
+            parent = article.parent_id
+            if parent:
+                article.child_ids.write({"parent_id": parent.id})
+        return super(Article, self).unlink()
+
+    #########
+    # Actions
+    #########
+
+    def action_home_page(self):
+        if 'res_id' not in self.env.context:
+            article = self.env['knowledge.article.favourite'].search([
+                ('user_id', '=', self.env.user.id),
+                ('article_id.user_has_access', '=', True)], limit=1).article_id
+            if not article:
+                article = self.search([
+                    ('parent_id', '=', False),
+                    ('internal_permission', '!=', 'none')
+                ], limit=1, order='sequence')
+                if not article:
+                    article = self.search([('parent_id', '=', False)], limit=1, order='sequence')
+        else:
+            article = self.browse(self.env.context['res_id'])
+        mode = 'edit' if article.user_can_write else 'readonly'
+        action = self.env['ir.actions.act_window']._for_xml_id('knowledge.knowledge_article_dashboard_action')
+        action['res_id'] = self.env.context.get('res_id', article.id)
+        action['context'] = dict(ast.literal_eval(action.get('context')), form_view_initial_mode=mode)
+        return action
+
+    def action_set_lock(self):
+        for article in self:
+            article.is_locked = True
+
+    def action_set_unlock(self):
+        for article in self:
+            article.is_locked = False
+
+    def action_toggle_favourite(self):
+        article = self.sudo()
+        if not article.user_has_access:
+            raise AccessError(_("You cannot add this article to your favourites"))
+        article.is_user_favourite = not article.is_user_favourite
+        return article.is_user_favourite
+
+    def action_archive(self):
+        return super(Article, self | self._get_descendants()).action_archive()
+
+    #####################
+    #  Business methods
+    #####################
+
+    def get_possible_parents(self, term=""):
+        # do a search_read and exclude all articles in descendants
+        exclude_ids = self._get_descendants()
+        exclude_ids |= self
+        return self.search_read(
+            domain=['&', ['name', '=ilike', '%%%s%%' % term], ['id', 'not in', exclude_ids.ids]],
+            fields=['id', 'icon', 'name'],
+        )
+
+    def move_to(self, parent_id=False, before_article_id=False, private=False):
```

Comment:
You should add docstrings and parameters explanation ... quite hard to understand what this does, have to play the guess game ^^ .